api/pkg/artifact/artifact.go

```diff
@@ -133,24 +178,120 @@ func (gac *GcsArtifactClient) DeleteArtifact(ctx context.Context, url string) er
 	return nil
 }
 
+type S3ArtifactClient struct {
+	URLScheme
+	Type   string
+	client *s3.Client
+}
+
+func NewS3ArtifactClient() (Service, error) {
+	cfg, err := config.LoadDefaultConfig(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("%s,failed loading s3 config for the artifact client", err.Error())
+	}
+	client := s3.NewFromConfig(cfg, func(o *s3.Options) {
+		o.BaseEndpoint = aws.String(os.Getenv("AWS_ENDPOINT_URL"))
+	})
+	return &S3ArtifactClient{
+		Type:      s3ArtifactClientType,
+		URLScheme: s3URLScheme,
+		client:    client,
+	}, nil
+}
+
+func (s3c *S3ArtifactClient) GetType() string {
+	return s3c.Type
+}
+
+func (s3c *S3ArtifactClient) ReadArtifact(ctx context.Context, url string) ([]byte, error) {
+	u, err := s3c.ParseURL(url)
+	if err != nil {
+		return nil, err
+	}
+
+	reader, err := s3c.client.GetObject(ctx, &s3.GetObjectInput{
+		Bucket: aws.String(u.Bucket),
+		Key:    aws.String(u.Object),
+	})
+	if err != nil {
+		var nsk *types.NoSuchKey
+		if errors.As(err, &nsk) {
+			return nil, ErrObjectNotExist
+		}
+		return nil, err
+	}
+	defer reader.Body.Close() //nolint:errcheck
+
+	bytes, err := io.ReadAll(reader.Body)
+	if err != nil {
+		return nil, err
+	}
+	return bytes, nil
+}
+
+func (s3c *S3ArtifactClient) WriteArtifact(ctx context.Context, url string, content []byte) error {
+	u, err := s3c.ParseURL(url)
+	if err != nil {
+		return err
+	}
+
+	_, err = s3c.client.PutObject(ctx, &s3.PutObjectInput{
+		Bucket: aws.String(u.Bucket),
+		Key:    aws.String(u.Object),
+		Body:   bytes.NewReader(content),
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s3c *S3ArtifactClient) DeleteArtifact(ctx context.Context, url string) error {
+	u, err := s3c.ParseURL(url)
+	if err != nil {
+		return err
+	}
+
+	// TODO: To confirm and refactor if versioning is enabled on S3 and to specify the versionId to be deleted
+	input := &s3.DeleteObjectInput{
+		Bucket: aws.String(u.Bucket),
+		Key:    aws.String(u.Object),
+	}
+
+	_, err = s3c.client.DeleteObject(ctx, input)
```

Comment:
What happens if you try to delete an object that doesn't exist? If it returns error should we return `ErrObjectNotExist` ?

Reply:
We can check if the error is of the `*types.NoSuchKey` type to see if the object doesn't exist but I just realised my current implementation doesn't actually delete anything because the `DeleteObject` can only delete object. I've refactored this method a little to now get all the objects specified within the Merlin artifact path (`mlflow/.../artifacts`) and then call the `DeleteObjects` method to make 1 single call to delete multiple objects at once.